From 245e063ee74b88479bd69246ba07bdfd52985ebd Mon Sep 17 00:00:00 2001
From: Job Noorman <jnoorman@igalia.com>
Date: Thu, 19 Dec 2024 17:46:46 +0100
Subject: [PATCH] ir3,freedreno: remove binning outputs after vs ucp lowering

nir_lower_clip_vs relies on VARYING_SLOT_CLIP_VERTEX which gets removed
when removing the outputs that are unused by the binning vs. Fix this by
only removing the outputs after running nir_lower_clip_vs.

Signed-off-by: Job Noorman <jnoorman@igalia.com>
Closes: https://gitlab.freedesktop.org/mesa/mesa/-/issues/12328
Fixes: 24b422dc3ec ("ir3: remove unused outputs for binning pass in NIR")
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/32727>
(cherry picked from commit 419879ee6292506c79a3f1342161663cac77b33c)
---
 .pick_status.json           |  2 +-
 src/freedreno/ir3/ir3_nir.c | 18 +++++++++---------
 2 files changed, 10 insertions(+), 10 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index 46763bf47cd4d..bfc21596fe8a1 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -694,7 +694,7 @@
         "description": "ir3,freedreno: remove binning outputs after vs ucp lowering",
         "nominated": true,
         "nomination_type": 2,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": "24b422dc3ec47b0a98a876acffeca258729a83e8",
         "notes": null
diff --git a/src/freedreno/ir3/ir3_nir.c b/src/freedreno/ir3/ir3_nir.c
index de6a38cfbb774..2dc9aa34f5c87 100644
--- a/src/freedreno/ir3/ir3_nir.c
+++ b/src/freedreno/ir3/ir3_nir.c
@@ -1049,15 +1049,6 @@ ir3_nir_lower_variant(struct ir3_shader_variant *so,
       }
    }
 
-   if (so->binning_pass) {
-      if (OPT(s, lower_binning)) {
-         progress = true;
-
-         /* outputs_written has changed. */
-         nir_shader_gather_info(s, nir_shader_get_entrypoint(s));
-      }
-   }
-
    /* Note that it is intentional to use the VS lowering pass for GS, since we
     * lower GS into something that looks more like a VS in ir3_nir_lower_gs():
     */
@@ -1068,6 +1059,15 @@ ir3_nir_lower_variant(struct ir3_shader_variant *so,
          progress |= OPT(s, nir_lower_clip_fs, so->key.ucp_enables, true);
    }
 
+   if (so->binning_pass) {
+      if (OPT(s, lower_binning)) {
+         progress = true;
+
+         /* outputs_written has changed. */
+         nir_shader_gather_info(s, nir_shader_get_entrypoint(s));
+      }
+   }
+
    /* Move large constant variables to the constants attached to the NIR
     * shader, which we will upload in the immediates range.  This generates
     * amuls, so we need to clean those up after.
-- 
GitLab

